qiskit_ibm_runtime/runtime_job_v2.py

```diff
@@ -0,0 +1,408 @@
+        decoder: Optional[Type[ResultDecoder]] = None,
+    ) -> Any:
+        """Return the results of the job.
+
+        Args:
+            timeout: Number of seconds to wait for job.
+            decoder: A :class:`ResultDecoder` subclass used to decode job results.
+
+        Returns:
+            Runtime job result.
+
+        Raises:
+            RuntimeJobFailureError: If the job failed.
+            RuntimeJobMaxTimeoutError: If the job does not complete within given timeout.
+            RuntimeInvalidStateError: If the job was cancelled, and attempting to retrieve result.
+        """
+        _decoder = decoder or self._final_result_decoder
+        self.wait_for_final_state(timeout=timeout)
+        if self._status == "ERROR":
+            error_message = self._reason if self._reason else self._error_message
+            if self._reason == "RAN TOO LONG":
+                raise RuntimeJobMaxTimeoutError(error_message)
+            raise RuntimeJobFailureError(f"Unable to retrieve job result. {error_message}")
+        if self._status == "CANCELLED":
+            raise RuntimeInvalidStateError(
+                "Unable to retrieve result for job {}. " "Job was cancelled.".format(self.job_id())
+            )
+
+        result_raw = self._download_external_result(
+            self._api_client.job_results(job_id=self.job_id())
+        )
+        # TODO: Remove getting/setting version once it's in result metadata
+        if _decoder.__name__ == EstimatorResultDecoder.__name__:
+            if not self._version:
+                self._version = self.inputs.get("version", 1)
+
+        return _decoder.decode(result_raw) if result_raw else None  # type: ignore
+
+    def cancel(self) -> None:
+        """Cancel the job.
+
+        Raises:
+            RuntimeInvalidStateError: If the job is in a state that cannot be cancelled.
+            IBMRuntimeError: If unable to cancel job.
+        """
+        try:
+            self._api_client.job_cancel(self.job_id())
+        except RequestsApiError as ex:
+            if ex.status_code == 409:
+                raise RuntimeInvalidStateError(f"Job cannot be cancelled: {ex}") from None
+            raise IBMRuntimeError(f"Failed to cancel job: {ex}") from None
+        self.cancel_result_streaming()
+        self._status = "CANCELLED"
+
+    def status(self) -> JobStatus:
+        """Return the status of the job.
+
+        Returns:
+            Status of this job.
+        """
+        self._set_status_and_error_message()
+        return self._status
+
+    def error_message(self) -> Optional[str]:
+        """Returns the reason if the job failed.
+
+        Returns:
+            Error message string or ``None``.
+        """
+        self._set_status_and_error_message()
+        return self._error_message
+
+    def _set_status_and_error_message(self) -> None:
+        """Fetch and set status and error message."""
+        if self._status not in JOB_FINAL_STATES:
```

Comment:
I don't think these methods need to be duplicated, especially since there are methods like `in_final_state()`

Reply:
If these status related methods are in `BaseRuntimeJob` instead, it makes things messier because `JOB_FINAL_STATES` is different, and all the statuses are strings instead of enums.

Reply:
But if the only difference is `JOB_FINAL_STATES`, and you're just checking if a status is in final state, can't you just call `if not self.in_final_state()` and have each child implement its own `in_final_state()` ?